crates/core/app/tests/common/test_node_builder_ext.rs

```diff
@@ -0,0 +1,120 @@
+pub trait BuilderExt: Sized {
+    /// The error thrown by [`with_penumbra_auto_app_state`]
+    type Error;
+    /// Add the provided Penumbra [`AppState`] to the builder.
+    ///
+    /// This will inject any configured validators into the state before serializing it into bytes.
+    fn with_penumbra_auto_app_state(self, app_state: AppState) -> Result<Self, Self::Error>;
+}
+
+impl BuilderExt for Builder {
+    type Error = anyhow::Error;
+    fn with_penumbra_auto_app_state(self, app_state: AppState) -> Result<Self, Self::Error> {
+        // Generate a penumbra validator using the test node's consensus keys (if they exist).
+        // Eventually, we may wish to generate and inject additional definitions, but only a single
+        // validator is supported for now.
+        let app_state = match self
+            .keys
+            .as_ref()
+            .map(generate_penumbra_validator)
+            .inspect(log_validator)
+            .map(std::iter::once)
+        {
+            Some(validator) => app_state_with_validators(app_state, validator)?,
+            None => app_state,
+        };
+
+        // Serialize the app state into bytes, and add it to the builder.
+        serde_json::to_vec(&app_state)
+            .map_err(Self::Error::from)
+            .map(|s| self.app_state(s))
+    }
+}
+
+/// Injects the given collection of [`Validator`s][PenumbraValidator] into the app state.
+fn app_state_with_validators<V>(
+    app_state: AppState,
+    validators: V,
+) -> Result<AppState, anyhow::Error>
+where
+    V: IntoIterator<Item = PenumbraValidator>,
+{
+    use AppState::{Checkpoint, Content};
+    match app_state {
+        Checkpoint(_) => anyhow::bail!("checkpoint app state isn't supported"),
+        Content(mut content) => {
+            // Inject the builder's validators into the staking component's genesis state.
+            std::mem::replace(
+                &mut content.stake_content.validators,
+                validators.into_iter().collect(),
+            )
+            .tap(|overwritten| {
+                // Log a warning if this overwrote any validators already in the app state.
+                if !overwritten.is_empty() {
+                    tracing::warn!(
+                        ?overwritten,
+                        "`with_penumbra_auto_app_state` overwrote validators in the given AppState"
+                    )
+                }
+            });
+            Ok(Content(content))
+        }
+    }
+}
+
+/// Generates a [`Validator`][PenumbraValidator] given a set of consensus [`Keys`].
+fn generate_penumbra_validator(
+    Keys {
+        consensus_verification_key,
+        ..
+    }: &Keys,
+) -> PenumbraValidator {
+    /// A temporary stub for validator keys.
+    ///
+    /// NB: for now, we will use the same key for governance. See the documentation of
+    /// `GovernanceKey` for more information about cold storage of validator keys.
+    const BYTES: [u8; 32] = [0; 32];
```

Comment:
This isn't a valid key, which may cause problems. 

We should instead generate a random signing key, get its verification key, use that for the identity key, and throw the signing key away after.

Reply:
i addressed this in 718fecff073986ab58a5de02ecc4f3676bef2124, in the form of some comments. i agree this will cause problems, but i am explicitly interested in having a failing test in hand before i add the identity key.